geoana/em/static/halfspace.py

```diff
@@ -0,0 +1,270 @@
+import numpy as np
+
+__all__ = [
+    "PointCurrentHalfSpace"
+]
+
+
+class PointCurrentHalfSpace:
+    """Class for a point current in a halfspace.
+
+    The ``PointCurrentHalfSpace`` class is used to analytically compute the
+    potentials, current densities and electric fields within a halfspace due to a point current.
+
+    Parameters
+    ----------
+    current : float
+        Electrical current in the point current (A). Default is 1A.
+    rho : float
+        Resistivity in the point current (:math:`\\Omega \\cdot m`).
+    location : array_like, optional
+        Location at which we are observing in 3D space (m). Default is (0, 0, 0).
+    """
+
+    def __init__(self, rho, current=1.0, location=None):
+
+        self.current = current
+        self.rho = rho
+        if location is None:
+            location = np.r_[0, 0, 0]
+        self.location = location
+
+    @property
+    def current(self):
+        """Current in the point current in Amps.
+
+        Returns
+        -------
+        float
+            Current in the point current in Amps.
+        """
+        return self._current
+
+    @current.setter
+    def current(self, value):
+
+        try:
+            value = float(value)
+        except:
+            raise TypeError(f"current must be a number, got {type(value)}")
+
+        if value <= 0.0:
+            raise ValueError("current must be greater than 0")
+
+        self._current = value
+
+    @property
+    def rho(self):
+        """Resistivity in the point current in :math:`\\Omega \\cdot m`.
+
+        Returns
+        -------
+        float
+            Resistivity in the point current in :math:`\\Omega \\cdot m`.
+        """
+        return self._rho
+
+    @rho.setter
+    def rho(self, value):
+
+        try:
+            value = float(value)
+        except:
+            raise TypeError(f"current must be a number, got {type(value)}")
+
+        if value <= 0.0:
+            raise ValueError("current must be greater than 0")
+
+        self._rho = value
+
+    @property
+    def location(self):
+        """Location of observer in 3D space.
+
+        Returns
+        -------
+        (3) numpy.ndarray of float
+            Location of observer in 3D space. Default = np.r_[0,0,0].
+        """
+        return self._location
+
+    @location.setter
+    def location(self, vec):
+
+        try:
+            vec = np.atleast_1d(vec).astype(float)
+        except:
+            raise TypeError(f"location must be array_like, got {type(vec)}")
+
+        if len(vec) != 3:
+            raise ValueError(
+                f"location must be array_like with shape (3,), got {len(vec)}"
+            )
+
```

Comment:
Should probably add a test here to make sure it' the `z <= 0` as the current source/sink can only exist in the halfspace.